File: rust/src/tools/claude.rs
//! Claude CLI tool configuration
//! Based on hive-mind's claude.lib.mjs implementation

use crate::streaming::parse_ndjson;
use serde_json::Value;
use std::collections::HashMap;

/// Get the Claude model map
pub fn get_model_map() -> HashMap<&'static str, &'static str> {
    let mut map = HashMap::new();
    map.insert("sonnet", "claude-sonnet-4-5-20250929");
    map.insert("opus", "claude-opus-4-5-20251101");
    map.insert("haiku", "claude-haiku-4-5-20251001");
    map.insert("haiku-3-5", "claude-3-5-haiku-20241022");
    map.insert("haiku-3", "claude-3-haiku-20240307");
    map
}

/// Map model alias to full model ID
///
/// # Arguments
/// * `model` - Model alias or full ID
///
/// # Returns
/// Full model ID
pub fn map_model_to_id(model: &str) -> String {
    let model_map = get_model_map();
    model_map
        .get(model)
        .map(|s| s.to_string())
        .unwrap_or_else(|| model.to_string())
}

/// Claude command build options
#[derive(Debug, Clone, Default)]
pub struct ClaudeBuildOptions {
    pub prompt: Option<String>,
    pub system_prompt: Option<String>,
    pub append_system_prompt: Option<String>,
    pub model: Option<String>,
    pub fallback_model: Option<String>,
    pub print: bool,
    pub verbose: bool,
    pub json: bool,
    pub json_input: bool,
    pub replay_user_messages: bool,
    pub resume: Option<String>,
    pub session_id: Option<String>,
    pub fork_session: bool,
    // Note: dangerously_skip_permissions is always enabled and not configurable (per issue #3)
}

impl ClaudeBuildOptions {
    /// Create new options
    pub fn new() -> Self {
        Self::default()
    }
}

/// Build command line arguments for Claude
///
/// # Arguments
/// * `options` - Build options
///
/// # Returns
/// Vector of CLI arguments
pub fn build_args(options: &ClaudeBuildOptions) -> Vec<String> {
    let mut args = Vec::new();

    // Permission bypass - always enabled, not configurable (per issue #3)
    args.push("--dangerously-skip-permissions".to_string());

    if let Some(ref model) = options.model {
        let mapped_model = map_model_to_id(model);
        args.push("--model".to_string());
        args.push(mapped_model);
    }

    if let Some(ref fallback_model) = options.fallback_model {
        let mapped_fallback = map_model_to_id(fallback_model);
        args.push("--fallback-model".to_string());
        args.push(mapped_fallback);
    }

    if let Some(ref prompt) = options.prompt {
        args.push("--prompt".to_string());
        args.push(prompt.clone());
    }

    if let Some(ref system_prompt) = options.system_prompt {
        args.push("--system-prompt".to_string());
        args.push(system_prompt.clone());
    }

    if let Some(ref append_system_prompt) = options.append_system_prompt {
        args.push("--append-system-prompt".to_string());
        args.push(append_system_prompt.clone());
    }

    if options.verbose {
        args.push("--verbose".to_string());
    }

    if options.print {
        args.push("-p".to_string()); // Print mode
    }

    // JSON output mode - use stream-json format per issue #3
    if options.json {
        args.push("--output-format".to_string());
        args.push("stream-json".to_string());
    }

    // JSON input mode - use stream-json format per issue #3
    if options.json_input {
        args.push("--input-format".to_string());
        args.push("stream-json".to_string());
    }

    // Replay user messages (only with stream-json input/output)
    if options.replay_user_messages {
        args.push("--replay-user-messages".to_string());
    }

    // Session management
    if let Some(ref session_id) = options.session_id {
        args.push("--session-id".to_string());
        args.push(session_id.clone());
    }

    if let Some(ref resume) = options.resume {
        args.push("--resume".to_string());
        args.push(resume.clone());
    }

    if options.fork_session {
        args.push("--fork-session".to_string());
    }

    args
}

/// Escape an argument for shell usage
fn escape_arg(arg: &str) -> String {
    if arg.contains('"')
        || arg.contains(char::is_whitespace)
        || arg.contains('$')
        || arg.contains('`')
        || arg.contains('\\')
    {
        let escaped = arg
            .replace('\\', "\\\\")
            .replace('"', "\\\"")
            .replace('$', "\\$")
            .replace('`', "\\`");
        format!("\"{}\"", escaped)
    } else {
        arg.to_string()
    }
}

/// Build complete command string for Claude
///
/// # Arguments
/// * `options` - Build options
///
/// # Returns
/// Complete command string
pub fn build_command(options: &ClaudeBuildOptions) -> String {
    let args = build_args(options);
    let args_str: Vec<String> = args.iter().map(|a| escape_arg(a)).collect();
    format!("claude {}", args_str.join(" ")).trim().to_string()
}

/// Parse JSON messages from Claude output
/// Claude outputs NDJSON (newline-delimited JSON) in JSON mode
///
/// # Arguments
/// * `output` - Raw output string
///
/// # Returns
/// Vector of parsed JSON messages
pub fn parse_output(output: &str) -> Vec<Value> {
    parse_ndjson(output)
}

/// Extract session ID from Claude output
///
/// # Arguments
/// * `output` - Raw output string
///
/// # Returns
/// Session ID or None
pub fn extract_session_id(output: &str) -> Option<String> {
    let messages = parse_output(output);

    for msg in messages {
        if let Some(session_id) = msg.get("session_id").and_then(|v| v.as_str()) {
            return Some(session_id.to_string());
        }
    }

    None
}

/// Usage statistics
#[derive(Debug, Clone, Default)]
pub struct ClaudeUsage {
    pub input_tokens: u64,
    pub output_tokens: u64,
    pub cache_creation_tokens: u64,
    pub cache_read_tokens: u64,
}

/// Extract usage statistics from Claude output
///
/// # Arguments
/// * `output` - Raw output string
///
/// # Returns
/// Usage statistics
pub fn extract_usage(output: &str) -> ClaudeUsage {
    let messages = parse_output(output);
    let mut usage = ClaudeUsage::default();

    for msg in messages {
        if let Some(msg_usage) = msg.get("message").and_then(|m| m.get("usage")) {
            if let Some(input) = msg_usage.get("input_tokens").and_then(|v| v.as_u64()) {
                usage.input_tokens += input;
            }
            if let Some(output) = msg_usage.get("output_tokens").and_then(|v| v.as_u64()) {
                usage.output_tokens += output;
            }
            if let Some(cache_creation) = msg_usage
                .get("cache_creation_input_tokens")
                .and_then(|v| v.as_u64())
            {
                usage.cache_creation_tokens += cache_creation;
            }
            if let Some(cache_read) = msg_usage
                .get("cache_read_input_tokens")
                .and_then(|v| v.as_u64())
            {
                usage.cache_read_tokens += cache_read;
            }
        }
    }

    usage
}

/// Claude tool configuration
#[derive(Debug, Clone)]
pub struct ClaudeTool {
    pub name: &'static str,
    pub display_name: &'static str,
    pub executable: &'static str,
    pub supports_json_output: bool,
    pub supports_json_input: bool,
    pub supports_system_prompt: bool,
    pub supports_append_system_prompt: bool,
    pub supports_resume: bool,
    pub supports_fork_session: bool,
    pub supports_session_id: bool,
    pub supports_fallback_model: bool,
    pub supports_verbose: bool,
    pub supports_replay_user_messages: bool,
    pub default_model: &'static str,
}

impl Default for ClaudeTool {
    fn default() -> Self {
        Self {
            name: "claude",
            display_name: "Claude Code CLI",
            executable: "claude",
            supports_json_output: true,
            supports_json_input: true, // Claude supports stream-json input format
            supports_system_prompt: true,
            supports_append_system_prompt: true, // Supports --append-system-prompt
            supports_resume: true,
            supports_fork_session: true,   // Supports --fork-session
            supports_session_id: true,     // Supports --session-id
            supports_fallback_model: true, // Supports --fallback-model
            supports_verbose: true,        // Supports --verbose
            supports_replay_user_messages: true, // Supports --replay-user-messages
            default_model: "sonnet",
        }
    }
}

// Tests are in rust/tests/claude_tests.rs

File: js/src/tools/claude.mjs
/**
 * Claude CLI tool configuration
 * Based on hive-mind's claude.lib.mjs implementation
 */

import { buildCommandHead, escapeArg, normalizeExtraArgs } from './shell.mjs';

/**
 * Available Claude model configurations
 * Maps aliases to full model IDs
 */
export const modelMap = {
  sonnet: 'claude-sonnet-4-6',
  opus: 'claude-opus-4-7', // Opus 4.7 (hive-mind issue #1620, PR #1621)
  haiku: 'claude-haiku-4-5-20251001',
  'haiku-3-5': 'claude-3-5-haiku-20241022',
  'haiku-3': 'claude-3-haiku-20240307',
  opusplan: 'opusplan', // Special mode: Opus for planning, Sonnet for execution
  // Shorter version aliases
  'sonnet-4-6': 'claude-sonnet-4-6',
  'opus-4-7': 'claude-opus-4-7',
  'opus-4-6': 'claude-opus-4-6',
  'opus-4-5': 'claude-opus-4-5-20251101',
  'sonnet-4-5': 'claude-sonnet-4-5-20250929', // Backward compatibility
  'haiku-4-5': 'claude-haiku-4-5-20251001',
  // Full model ID aliases for backward compatibility
  'claude-sonnet-4-6': 'claude-sonnet-4-6',
  'claude-opus-4-7': 'claude-opus-4-7',
  'claude-opus-4-6': 'claude-opus-4-6',
  'claude-opus-4-5': 'claude-opus-4-5-20251101',
  'claude-sonnet-4-5': 'claude-sonnet-4-5-20250929',
  'claude-haiku-4-5': 'claude-haiku-4-5-20251001',
};

/**
 * Map model alias to full model ID
 * @param {Object} options - Options
 * @param {string} options.model - Model alias or full ID
 * @returns {string} Full model ID
 */
export function mapModelToId(options) {
  const { model } = options;
  return modelMap[model] || model;
}

/**
 * Build command line arguments for Claude
 * @param {Object} options - Options
 * @param {string} [options.prompt] - User prompt
 * @param {string} [options.systemPrompt] - System prompt
 * @param {string} [options.appendSystemPrompt] - System prompt to append to default
 * @param {string} [options.model] - Model to use
 * @param {string} [options.fallbackModel] - Fallback model when default is overloaded
 * @param {boolean} [options.print] - Print mode (non-interactive)
 * @param {boolean} [options.verbose] - Verbose mode
 * @param {boolean} [options.json] - JSON output mode (stream-json format)
 * @param {boolean} [options.jsonInput] - JSON input mode (stream-json format)
 * @param {boolean} [options.replayUserMessages] - Re-emit user messages on stdout
 * @param {string} [options.resume] - Resume session ID
 * @param {string} [options.sessionId] - Use specific session ID (must be valid UUID)
 * @param {boolean} [options.forkSession] - Create new session ID when resuming
 * @param {boolean} [options.readOnly] - Use Claude plan permission mode
 * @param {string[]} [options.extraArgs] - Extra raw CLI args appended after typed args
 * @param {boolean} [options.skipDefaultSafetyFlags] - Do not add default bypass flags
 * @param {string} [options.permissionMode] - Explicit Claude permission mode
 * @returns {string[]} Array of CLI arguments
 */
export function buildArgs(options) {
  const {
    prompt,
    systemPrompt,
    appendSystemPrompt,
    model,
    fallbackModel,
    print = false,
    verbose = false,
    json = false,
    jsonInput = false,
    replayUserMessages = false,
    resume,
    sessionId,
    forkSession = false,
    readOnly = false,
    extraArgs = [],
    skipDefaultSafetyFlags = false,
    permissionMode,
  } = options;

  const args = [];

  if (readOnly) {
    args.push('--permission-mode', 'plan');
  } else if (permissionMode) {
    args.push('--permission-mode', permissionMode);
  } else if (!skipDefaultSafetyFlags) {
    // Permission bypass - enabled by default for autonomous execution
    args.push('--dangerously-skip-permissions');
  }

  if (model) {
    const mappedModel = mapModelToId({ model });
    args.push('--model', mappedModel);
  }

  if (fallbackModel) {
    const mappedFallback = mapModelToId({ model: fallbackModel });
    args.push('--fallback-model', mappedFallback);
  }

  if (prompt) {
    args.push('--prompt', prompt);
  }

  if (systemPrompt) {
    args.push('--system-prompt', systemPrompt);
  }

  if (appendSystemPrompt) {
    args.push('--append-system-prompt', appendSystemPrompt);
  }

  if (verbose) {
    args.push('--verbose');
  }

  if (print) {
    args.push('-p'); // Print mode
  }

  // JSON output mode - use stream-json format per issue #3
  if (json) {
    args.push('--output-format', 'stream-json');
  }

  // JSON input mode - use stream-json format per issue #3
  if (jsonInput) {
    args.push('--input-format', 'stream-json');
  }

  // Replay user messages (only with stream-json input/output)
  if (replayUserMessages) {
    args.push('--replay-user-messages');
  }

  // Session management
  if (sessionId) {
    args.push('--session-id', sessionId);
  }

  if (resume) {
    args.push('--resume', resume);
  }

  if (forkSession) {
    args.push('--fork-session');
  }

  args.push(...normalizeExtraArgs(extraArgs));

  return args;
}

/**
 * Build complete command string for Claude
 * @param {Object} options - Options
 * @param {string} options.workingDirectory - Working directory
 * @param {string} [options.prompt] - User prompt
 * @param {string} [options.promptFile] - File containing prompt input
 * @param {string} [options.systemPrompt] - System prompt
 * @param {string} [options.appendSystemPrompt] - System prompt to append to default
 * @param {string} [options.model] - Model to use
 * @param {string} [options.fallbackModel] - Fallback model when default is overloaded
 * @param {boolean} [options.print] - Print mode (non-interactive)
 * @param {boolean} [options.verbose] - Verbose mode
 * @param {boolean} [options.json] - JSON output mode (stream-json format)
 * @param {boolean} [options.jsonInput] - JSON input mode (stream-json format)
 * @param {boolean} [options.replayUserMessages] - Re-emit user messages on stdout
 * @param {string} [options.resume] - Resume session ID
 * @param {string} [options.sessionId] - Use specific session ID (must be valid UUID)
 * @param {boolean} [options.forkSession] - Create new session ID when resuming
 * @param {boolean} [options.readOnly] - Use Claude plan permission mode
 * @param {string} [options.executable='claude'] - Executable path/name
 * @param {Object|Array} [options.extraEnv] - Environment variables for the tool
 * @param {string[]} [options.extraArgs] - Extra raw CLI args appended after typed args
 * @param {boolean} [options.skipDefaultSafetyFlags] - Do not add default bypass flags
 * @param {string} [options.permissionMode] - Explicit Claude permission mode
 * @returns {string} Complete command string
 */
export function buildCommand(options) {
  const {
    prompt,
    promptFile,
    executable = 'claude',
    extraEnv,
    ...argOptions
  } = options;
  const args = buildArgs({
    ...argOptions,
    prompt: promptFile ? undefined : prompt,
  });
  const command =
    `${buildCommandHead({ executable, extraEnv })} ${args.map(escapeArg).join(' ')}`.trim();

  if (promptFile) {
    return `cat ${escapeArg(promptFile)} | ${command}`;
  }

  return command;
}

/**
 * Parse JSON messages from Claude output
 * Claude outputs NDJSON (newline-delimited JSON) in JSON mode
 * @param {Object} options - Options
 * @param {string} options.output - Raw output string
 * @returns {Object[]} Array of parsed JSON messages
 */
export function parseOutput(options) {
  const { output } = options;
  const messages = [];
  const lines = output.split('\n');

  for (const line of lines) {
    const trimmed = line.trim();
    if (!trimmed || !trimmed.startsWith('{')) {
      continue;
    }

    try {
      const parsed = JSON.parse(trimmed);
      messages.push(parsed);
    } catch {
      // Skip lines that aren't valid JSON
    }
  }

  return messages;
}

/**
 * Extract session ID from Claude output
 * @param {Object} options - Options
 * @param {string} options.output - Raw output string
 * @returns {string|null} Session ID or null
 */
export function extractSessionId(options) {
  const { output } = options;
  const messages = parseOutput({ output });

  for (const msg of messages) {
    if (msg.session_id) {
      return msg.session_id;
    }
  }

  return null;
}

/**
 * Extract usage statistics from Claude output
 * @param {Object} options - Options
 * @param {string} options.output - Raw output string
 * @returns {Object} Usage statistics
 */
export function extractUsage(options) {
  const { output } = options;
  const messages = parseOutput({ output });

  const usage = {
    inputTokens: 0,
    outputTokens: 0,
    cacheCreationTokens: 0,
    cacheReadTokens: 0,
  };

  for (const msg of messages) {
    if (msg.message?.usage) {
      const u = msg.message.usage;
      if (u.input_tokens) {
        usage.inputTokens += u.input_tokens;
      }
      if (u.output_tokens) {
        usage.outputTokens += u.output_tokens;
      }
      if (u.cache_creation_input_tokens) {
        usage.cacheCreationTokens += u.cache_creation_input_tokens;
      }
      if (u.cache_read_input_tokens) {
        usage.cacheReadTokens += u.cache_read_input_tokens;
      }
    }
  }

  return usage;
}

/**
 * Claude tool configuration
 */
export const claudeTool = {
  name: 'claude',
  displayName: 'Claude Code CLI',
  executable: 'claude',
  supportsJsonOutput: true,
  supportsJsonInput: true, // Claude supports stream-json input format
  supportsSystemPrompt: true,
  supportsAppendSystemPrompt: true, // Supports --append-system-prompt
  supportsResume: true,
  supportsForkSession: true, // Supports --fork-session
  supportsSessionId: true, // Supports --session-id
  supportsFallbackModel: true, // Supports --fallback-model
  supportsVerbose: true, // Supports --verbose
  supportsReplayUserMessages: true, // Supports --replay-user-messages
  supportsReadOnly: true, // Supports --permission-mode plan
  defaultModel: 'sonnet',
  modelMap,
  mapModelToId,
  buildArgs,
  buildCommand,
  parseOutput,
  extractSessionId,
  extractUsage,
};
